File: backend/app/api/v1/job.py
from flask import Blueprint, jsonify, request as flask_request
from sqlalchemy import func
import requests as http_requests
import traceback

from app.extensions import db
from app.models.interview import Interview
from app.models.job import Job
from app.models.user import User

DEVTO_ARTICLE_API = 'https://dev.to/api/articles/'


job_bp = Blueprint('job', __name__)


def success_response(data=None, msg='success', status_code=200):
    return jsonify({"code": 200, "data": data, "msg": msg, "message": msg}), status_code


@job_bp.route('', methods=['GET'])
def list_jobs():
    jobs = Job.query.order_by(Job.id.asc()).all()
    result = [job.to_dict() for job in jobs]
    return success_response(result)


@job_bp.route('/popular', methods=['GET'])
def list_popular_jobs():
    rows = db.session.query(
        Job,
        func.count(User.id).label('selected_count')
    ).join(
        User, User.default_job_id == Job.id
    ).group_by(
        Job.id
    ).order_by(
        func.count(User.id).desc(),
        Job.id.asc()
    ).limit(5).all()

    result = []
    for job, selected_count in rows:
        item = job.to_dict()
        item['selected_count'] = int(selected_count or 0)
        result.append(item)

    return success_response(result)


@job_bp.route('/avg-scores', methods=['GET'])
def list_job_average_scores():
    rows = db.session.query(
        Job,
        func.avg(Interview.total_score).label('avg_score'),
        func.count(Interview.id).label('interview_count'),
        func.count(func.distinct(Interview.user_id)).label('user_count')
    ).outerjoin(
        Interview,
        (Interview.job_id == Job.id)
        & (Interview.status == 'completed')
        & (Interview.total_score.isnot(None))
    ).group_by(
        Job.id
    ).order_by(
        Job.id.asc()
    ).all()

    result = []
    for job, avg_score, interview_count, user_count in rows:
        item = job.to_dict()
        item['avg_score'] = int(round(float(avg_score))) if avg_score is not None else 0
        item['interview_count'] = int(interview_count or 0)
        item['user_count'] = int(user_count or 0)
        result.append(item)

    return success_response(result)


# =============================================
# 技术热榜聚合 —— 多平台公开 API，实时获取技术社区热门话题
# =============================================
from app.services.trending_service import TrendingService

@job_bp.route('/trending', methods=['GET'])
def get_trending_topics():
    """
    多平台聚合技术热榜：掘金（多分类）+ Dev.to
    """
    job_key = flask_request.args.get('jobId', 'default')
    limit = min(int(flask_request.args.get('limit', 9)), 20)
    return success_response(TrendingService.get_trending_topics(job_key, limit))


@job_bp.route('/trending/detail', methods=['GET'])
def get_article_detail():
    """
    获取文章全文内容（目前支持 Dev.to 源）
    """
    article_id = flask_request.args.get('id', '')
    source = flask_request.args.get('source', '')

    if source == 'devto' and article_id.startswith('devto_'):
        devto_id = article_id.replace('devto_', '')
        try:
            resp = http_requests.get(
                DEVTO_ARTICLE_API + devto_id,
                headers={'User-Agent': 'AI-Interview-Platform/1.0'},
                timeout=8
            )
            resp.raise_for_status()
            data = resp.json()
            return success_response({
                'id': article_id,
                'title': data.get('title', ''),
                'body_markdown': data.get('body_markdown', ''),
                'body_html': data.get('body_html', ''),
                'url': data.get('url', ''),
                'tags': data.get('tag_list', []),
                'cover_image': data.get('cover_image', ''),
                'user': data.get('user', {}).get('name', ''),
                'reading_time': data.get('reading_time_minutes', 0),
                'published_at': data.get('published_at', ''),
            })
        except Exception:
            traceback.print_exc()
            return success_response(None, msg='文章获取失败')

    # 掘金文章无法获取全文，返回提示
    return success_response({
        'id': article_id,
        'title': '',
        'body_html': '',
        'body_markdown': '',
        'fallback': True,
        'msg': '该来源暂不支持全文阅读，请点击原文链接查看'
    })

File: frontend/src/api/job.js
// =============================================
// frontend/src/api/job.js
// 职位相关API
// Mock数据支持职位列表展示，对接后端时替换为真实request调用即可
// =============================================

import request from '@/utils/request'
const USE_MOCK = process.env.VUE_APP_USE_MOCK !== 'false'

export const fetchJobs = async () => {
  if (USE_MOCK) return null  // Mock 模式下由 constants.js 兜底
  const res = await request.get('/jobs')
  return res  // 返回数组：[{ id:1, name:'Java后端开发', description, tech_stack, icon_url }]
}

/**
 * 获取岗位均分（用于前端展示）
 * 返回格式：[{ id, avg_score, interview_count, user_count }, ...]
 */
export const fetchJobAvgScores = async () => {
  if (USE_MOCK) return []
  const res = await request.get('/jobs/avg-scores')
  return res
}

/**
 * 获取热门岗位（按用户默认选择数排名）
 * 返回数组与 /jobs 接口结构相同，前端仅需要 id 和 name
 */
export const fetchPopularJobs = async () => {
  if (USE_MOCK) return []
  const res = await request.get('/jobs/popular')
  return res
}

/**
 * 获取多平台聚合技术热榜（掘金多分类 + Dev.to）
 * @param {string} jobId - 岗位key
 * @param {number} limit - 返回条数
 * @returns {Promise<Array>} 热榜文章列表
 */
export const getTrendingTopics = async (jobId = 'default', limit = 9) => {
  return request.get('/jobs/trending', { params: { jobId, limit } })
}

/**
 * 获取文章全文内容（Dev.to 源支持站内阅读）
 * @param {string} id - 文章 ID
 * @param {string} source - 来源平台
 * @returns {Promise<Object>} 包含 body_html / body_markdown 的文章详情
 */
export const getArticleDetail = async (id, source) => {
  return request.get('/jobs/trending/detail', { params: { id, source } })
}
